Author: Alastair McKinstry <mckinstry@debian.org>
Bug-Upstream: https://github.com/open-mpi/ompi/pull/5537
Description: Fix for hang between ORTE / PMIX layers
Last-Updated: 2018-08-19
Forwarded: not-needed

Index: openmpi-3.1.1.real/orte/orted/pmix/pmix_server_gen.c
===================================================================
--- openmpi-3.1.1.real.orig/orte/orted/pmix/pmix_server_gen.c
+++ openmpi-3.1.1.real/orte/orted/pmix/pmix_server_gen.c
@@ -859,6 +859,15 @@ void pmix_tool_connected_fn(opal_list_t
 
 }
 
+static void lgcbfn(int sd, short args, void *cbdata)
+{
+    orte_pmix_server_op_caddy_t *cd = (orte_pmix_server_op_caddy_t*)cbdata;
+    if (NULL != cd->cbfunc) {
+        cd->cbfunc(cd->status, cd->cbdata);
+    }
+    OBJ_RELEASE(cd);
+}
+
 void pmix_server_log_fn(opal_process_name_t *requestor,
                         opal_list_t *info,
                         opal_list_t *directives,
@@ -907,6 +916,14 @@ void pmix_server_log_fn(opal_process_nam
     if (NULL != cbfunc) {
         cbfunc(OPAL_SUCCESS, cbdata);
     }
+
+    /* we cannot directly execute the callback here
+     * as it would threadlock - so shift to somewhere
+     * safe */
+    rc = ORTE_SUCCESS;  // unused - silence compiler warning
+    ORTE_PMIX_THREADSHIFT(requestor, NULL, rc,
+                          NULL, NULL, lgcbfn,
+                          cbfunc, cbdata);
 }
 
 int pmix_server_job_ctrl_fn(const opal_process_name_t *requestor,
